File: adapters/utility-adapter/go/adapter/security.go
package adapter
import ("crypto/hmac"; "crypto/sha256"; "encoding/hex"; "math"; "strconv"; "strings"; "time")
func VerifySignature(secret string, timestamp string, signature string, body []byte, window int64) bool {
    ts, err := strconv.ParseInt(timestamp, 10, 64)
    if err != nil || math.Abs(float64(time.Now().Unix()-ts)) > float64(window) { return false }
    mac := hmac.New(sha256.New, []byte(secret))
    mac.Write([]byte(timestamp + "."))
    mac.Write(body)
    expected := hex.EncodeToString(mac.Sum(nil))
    provided := strings.TrimPrefix(signature, "v1=")
    return hmac.Equal([]byte(expected), []byte(provided))
}

File: adapters/utility-adapter/python/efn_utility_adapter/security.py
import hmac
import hashlib
import time

def verify_signature(secret: bytes, timestamp: str, signature: str, body: bytes, window: int = 300) -> bool:
    try:
        if abs(time.time() - int(timestamp)) > window:
            return False
        expected = hmac.new(secret, timestamp.encode() + b'.' + body, hashlib.sha256).hexdigest()
        provided = signature.replace('v1=', '')
        return hmac.compare_digest(expected, provided)
    except Exception:
        return False
